spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/MergeInto.scala

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.spark.sql.catalyst.plans.logical
+
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.expressions.codegen.GeneratePredicate
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation
+import org.apache.spark.sql.functions.col
+
+case class MergeInto(mergeIntoProcessor: MergeIntoProcessor,
+                     targetRelation: DataSourceV2Relation,
+                     child: LogicalPlan) extends UnaryNode {
+  override def output: Seq[Attribute] = targetRelation.output
+}
+
+class MergeIntoProcessor(isSourceRowNotPresent: Expression,
+                         isTargetRowNotPresent: Expression,
+                         matchedConditions: Seq[Expression],
+                         matchedOutputs: Seq[Seq[Expression]],
+                         notMatchedConditions: Seq[Expression],
+                         notMatchedOutputs: Seq[Seq[Expression]],
+                         targetOutput: Seq[Expression],
+                         joinedAttributes: Seq[Attribute]) extends Serializable {
```

Comment:
This is essentially a physical plan node that is linked into both the physical plan and logical plan. I think it should be a normal physical plan node that is created in a strategy, just like other plans.

The main issue with the way this PR currently works is that it doesn't delegate enough to the rest of the Spark planner. All of the analysis is done during rewrite in the optimizer, for example. I think that this should be broken up into analysis rules to validate and update the `MergeInto` plan, the rewrite rule to build the optimizations and join, and a strategy to convert the logical plan into a `MergeIntoExec`. I think this should also have a validation rule that checks each action to ensure that the expressions for that action are correctly resolved.

Reply:
@rdblue Can you please explain the idea bit more, specifically the `should be broken up into analysis rules to validate and update `MergeInto` plan` . Currently, we produce the `MergeInto` logical plan in the optimizer phase ? So we have gone past analysis at this point right ? The input SQL has already been parsed and resolved using `MergeIntoTable` by spark at this point i.e all the `mergeinto` inputs have been resolved ?

Reply:
We need to make sure there are analysis rules that guarantee the assumptions in this class. One possible issue that jumped out to both @aokolnychyi and I was that this assumes the expressions for insert and update actions are correct for the output of this node. We need to make sure that is the case.

Originally, I asked on Slack how that validation was being done, but I saw Anton's comment about it and I thought that probably meant that it isn't being done. If there are already rules in Spark to resolve and validate the plan, then that's great but we need to identify them and make a note here that we're relying on those for correctness. I still suspect that there aren't rules in Spark doing this because this is running the analyzer on expressions.

Reply:
@rdblue Sorry Ryan. I didn't notice your comment on slack until now. So currently Spark's Analyzer ensures that the `ResolveIntoTable` is fully resolved. 

[code](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L1634-L1672)

However, you are right that we don't do any semantics analysis on the plan currently. We should add it.

Reply:
Thanks for pointing me to the code! Looks like I was looking into it at the time you were writing this, which is why my comment below was just a bit later. I think we're all on the same page now.